pallets/foreign-investments/src/lib.rs

```diff
@@ -422,35 +302,473 @@ pub mod pallet {
+			kind: ForeignSwapKind,
+			swap_id: T::SwapId,
+		) {
+			ForeignSwapIdToSwapId::<T>::remove(&who, (investment_id, kind));
+			SwapidToForeignSwapId::<T>::remove(swap_id);
+		}
+
+		/// Returns the `amount_out` of the swap, seeing as the amount used to
+		/// swap.
+		fn amount_used_to_swap(swap: &SwapOf<T>) -> Result<T::Balance, DispatchError> {
+			T::CurrencyConverter::stable_to_stable(
+				swap.currency_out,
+				swap.currency_in,
+				swap.amount_in,
+			)
+		}
+
+		/// Apply a swap over a current possible swap state.
+		/// - If there was no previous swap, it adds it.
+		/// - If there was a swap in the same direction, it increments it.
+		/// - If there was a swap in the opposite direction:
+		///   - If the amount is smaller, it decrements it.
+		///   - If the amount is the same, it removes the inverse swap.
+		///   - If the amount is greater, it removes the inverse swap and create
+		///     another with the excess
+		///
+		/// The returned status contains the swapped amount after this call and
+		/// the pending amount to be swapped.
+		fn apply_swap(
+			who: &T::AccountId,
+			investment_id: T::InvestmentId,
+			kind: ForeignSwapKind,
+			new_swap: SwapOf<T>,
+		) -> Result<SwapStatus<T>, DispatchError> {
+			match ForeignSwapIdToSwapId::<T>::get(&who, (investment_id, kind)) {
+				None => {
+					let swap_id = T::TokenSwaps::place_order(
+						who.clone(),
+						new_swap.currency_in,
+						new_swap.currency_out,
+						new_swap.amount_in,
+						T::BalanceRatio::one(),
+					)?;
+					Self::register_swap(&who, investment_id, kind, swap_id);
+
+					Ok(SwapStatus {
+						swapped: T::Balance::zero(),
+						pending: new_swap.amount_in,
+					})
+				}
+				Some(swap_id) => {
+					let swap = T::TokenSwaps::get_order_details(swap_id)
+						.ok_or(Error::<T>::SwapOrderNotFound)?;
+
+					if swap.is_same_direction(&new_swap)? {
+						T::TokenSwaps::update_order(
+							who.clone(),
+							swap_id,
+							swap.amount_in.ensure_add(new_swap.amount_in)?,
+							T::BalanceRatio::one(),
+						)?;
+
+						Ok(SwapStatus {
+							swapped: T::Balance::zero(),
+							pending: new_swap.amount_in,
+						})
+					} else {
+						let inverse_swap = swap;
+						let new_swap_amount_out = Self::amount_used_to_swap(&new_swap)?;
+
+						match inverse_swap.amount_in.cmp(&new_swap_amount_out) {
+							Ordering::Less => {
+								T::TokenSwaps::update_order(
+									who.clone(),
+									swap_id,
+									inverse_swap.amount_in.ensure_sub(new_swap_amount_out)?,
+									T::BalanceRatio::one(),
+								)?;
+
+								Ok(SwapStatus {
+									swapped: new_swap.amount_in,
+									pending: T::Balance::zero(),
+								})
+							}
+							Ordering::Equal => {
+								T::TokenSwaps::cancel_order(swap_id.clone())?;
+								Self::unregister_swap(&who, investment_id, kind, swap_id);
+
+								Ok(SwapStatus {
+									swapped: new_swap.amount_in,
+									pending: T::Balance::zero(),
+								})
```

Comment:
Aren't we forgetting to notify the other part of the swap about the update here? the new submission gets a status, but the first submission never gets notified as the order is just canceled.

Reply:
That's a really good point 👏🏻 . I think you're right. I would need to think about a clean solution for it. It seems not obvious.

Definitely, moving this to `order-book` will be the simplest solution. `order-book` will call the corresponding callbacks